Point_set_processing_3/include/CGAL/scanline_orient_normals.h

```diff
@@ -0,0 +1,552 @@
+    std::tie (scan_position, solver_success)
+      = estimate_scan_position (begin, end, point_map, lines_of_sight);
+
+#ifdef CGAL_SCANLINE_ORIENT_VERBOSE
+    if (solver_success && scan_position.z() > mean_z)
+      std::cerr << "SOLVED" << std::endl;
+    else if (!solver_success)
+      std::cerr << "FAILED, solver failure" << std::endl;
+    else
+      std::cerr << "FAILED, scanner under scanline" << std::endl;
+#endif
+  }
+
+  std::size_t idx = 0;
+  for (Iterator it = begin; it != end; ++ it, ++ idx)
+  {
+    const Vector_3 los = lines_of_sight[idx];
+    const Vector_3& normal = get (normal_map, *it);
+    if (normal_along_scanline_is_inverted (normal, los))
+      put (normal_map, *it, -normal);
+  }
+}
+
+template <typename Iterator, typename PointMap, typename NormalMap,
+          typename ScanAngleMap>
+void orient_scanline (Iterator begin, Iterator end,
+                      PointMap point_map,
+                      NormalMap normal_map,
+                      ScanAngleMap,
+                      const Tag_true&) // fallback scan angle
+{
+  using Point_3 = typename boost::property_traits<PointMap>::value_type;
+  using Vector_3 = typename boost::property_traits<NormalMap>::value_type;
+
+  Vector_3 direction;
+  Vector_3 vertical;
+  std::tie (direction, vertical)
+    = scanline_base (begin, end, point_map);
+
+  // Estimate scanner position:
+  // average XY-projected point, located above
+  double mean_x = 0.;
+  double mean_y = 0.;
+  double max_z = -(std::numeric_limits<double>::max)();
+  std::size_t nb = 0;
+
+  for (Iterator it = begin; it != end; ++ it)
+  {
+    const Point_3& p = get (point_map, *it);
+    mean_x += p.x();
+    mean_y += p.y();
+    max_z = (std::max(max_z, p.z()));
+    ++ nb;
+  }
+
+  Iterator last = end; -- last;
+  double length = CGAL::approximate_sqrt (CGAL::squared_distance
+                                          (get (point_map, *begin),
+                                           get (point_map, *last)));
+
+  Point_3 scan_position (mean_x / nb, mean_y / nb,
+                         max_z + length * 2);
+
+  for (Iterator it = begin; it != end; ++ it)
+  {
+    Vector_3 line_of_sight (get(point_map, *it), scan_position);
+    const Vector_3& normal = get (normal_map, *it);
+    if (normal_along_scanline_is_inverted (normal, line_of_sight))
+      put (normal_map, *it, -normal);
+  }
+}
+
+} // namespace internal
+
+} // namespace Point_set_processing_3
+/// \endcond
+
+// ----------------------------------------------------------------------------
+// Public section
+// ----------------------------------------------------------------------------
+
+/**
+   \ingroup PkgPointSetProcessing3Algorithms
+
+   Orients the normals of the range of `points` by estimating a line
+   of sight and checking its consistency with the current normal orientation.
+
+   \warning this function requires the input `points` to be ordered
+   along scanlines aligned on the XY-plane. It is typically designed
```

Comment:
does XY matter? what if I scan a building facade?

Reply:
The algorithm is mainly designed for 2.5D point set acquired by airborne sensors.

If you scan a building facade, allowing random planes as reference won't solve the issue as your scanlines are not even guaranteed to be 2.5D (most ground sensors will scan all around them, not on a single plane). In that case, this algorithm is just not adapted.